frontend/public/components/utils/resource-log.jsx

```diff
@@ -0,0 +1,255 @@
+    </span>
+    { props.name || <LoadingInline /> }
+  </span>;
+};
+
+// Component to render resource dropdown or just the resource name if only
+// one resource is provided.
+const ContainerSelect = (props) => {
+  const currentResourceNameWithIcon = <NameWithIcon name={props.currentResource} kind={props.kind} />;
+
+  if (props.selectableResources.length > 1) {
+    const dropdownItems = _.map(props.selectableResources, (name) => {
+      return <NameWithIcon name={name} kind={props.kind} />;
+    });
+    return <Dropdown className="btn-group" items={dropdownItems} title={currentResourceNameWithIcon} onChange={props.selectResource} />;
+  }
+  return currentResourceNameWithIcon;
+};
+
+// Component for the streaming controls and info displayed above the log window
+const LogControls = (props) => {
+  return <div className="co-m-pane__top-controls">
+    { props.logState === 'loading' && <span className="co-icon-space-l"><LoadingInline /></span> }
+    { ['streaming', 'paused', 'eof'].includes(props.logState) && <TogglePlay active={props.logState === 'streaming'} onClick={props.toggleStreaming} disabled={props.logState === 'eof'}/> }
+    <span className="log-container-selector__text">
+      { props.logState === 'loading' && <span>Loading log...</span> }
+      { props.logState === 'streaming' && <span>Streaming logs from</span> }
+      { props.logState === 'paused' && <span>Log stream is paused.</span> }
+      { props.logState === 'eof' && <span>Log stream ended from </span> }
+    </span>
+    <ContainerSelect {...props} />
+  </div>;
+};
+
+export class ResourceLog extends SafetyFirst {
+  constructor(props) {
+    super(props);
+
+    this._buffer = lineBuffer(1000);
+    this._loadTime = Date.now();
+    this._pendingReload = null;
+
+    this._touchLoadTimeState = _.throttle(this._touchLoadTimeState, 100);
+    this._updateLogState = this._updateLogState.bind(this);
+    this._toggleStreaming = this._toggleStreaming.bind(this);
+
+    this.state = {
+      logURL: '',
+      loadTime: 0,
+      logState: 'loading'
+    };
+
+    this.state = _.defaults({}, this._initialState(), this.state);
+  }
+
+  componentDidMount() {
+    super.componentDidMount();
+    this._beginStreaming();
+  }
+
+  componentWillReceiveProps(nextProps) {
+    // If container changed, restart streaming
+    if (nextProps.container !== this.props.container) {
+      this._restartStreaming(nextProps);
+    } else {
+      this.setState(this._initialState(nextProps));
+    }
+  }
+
+  componentWillUnmount() {
+    super.componentWillUnmount();
+    this._endStreaming();
+  }
+
+  _initialState(props = this.props) {
+    const newState = {};
+    const namespace = _.get(props.obj, 'metadata.namespace');
+    const objName = _.get(props.obj, 'metadata.name');
+
+    // If the container changed, reset log state to loading, otherwise keep it as is
+    const initialLogState = this.props.container !== props.container ? 'loading' : this.state.logState;
```

Comment:
> there is still a bug I've been trying to figure out. The first time visiting a log view in either pods or builds, it is initially stuck in the 'loading' state and I can't quite figure out why. On a page reload or forcing a state change by scrolling the log, it updates, and everything works as expected

This is probably causing loading error you mentioned, did you mean to check <code>this.props. currentResource</code> in here?

Reply:
I think this function is going to go away. I'm refactoring now to use`getDerivedStatFromProps` so that this function won't be necessary.

Reply:
And yes, I also caught the missed `container` reference yesterday. It will be fixed in my next commit.